appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,725 @@
+};
+
+/**
+ * Queues life simulation for mouse.
+ * @param {Genetics.Mouse} mouse The mouse to queue the simulation for.
+ */
+Genetics.Cage.life = function(mouse) {
+  // Queue life simulation for mouse with some variation in timeout to allow for
+  // variation in order.
+  Genetics.Cage.lifePidsMap_[mouse.id] =
+      setTimeout(goog.partial(Genetics.Cage.simulateLife, mouse),
+          10 * Math.random());
+};
+
+/**
+ * Runs life simulation on mouse.
+ * @param {Genetics.Mouse} mouse The mouse to run the simulation for.
+ */
+Genetics.Cage.simulateLife = function(mouse) {
+  // Remove life from mapping of lives that are queued to run.
+  delete Genetics.Cage.lifePidsMap_[mouse.id];
+  // Age mouse.
+  mouse.age++;
+  if (mouse.aggressiveness > 0) {
+    // Mouse has fight attempts left, try to start a fight with another mouse
+    Genetics.Cage.instigateFight(mouse);
+  } else if (mouse.fertility > 0) {
+    // Mouse is still fertile, try to breed with another mouse.
+    Genetics.Cage.tryMate(mouse);
+  } else {
+    // Mouse has completed life cycle and dies.
+    new Genetics.Cage.Event('RETIRE', mouse.id).addToQueue();
+    Genetics.Cage.death(mouse);
+  }
+  // Queue life simulation again if mouse is still alive.
+  if (Genetics.Cage.isAlive(mouse)) {
+    Genetics.Cage.life(mouse);
+  }
+  // Check for end game.
+  Genetics.Cage.checkForEnd();
+};
+
+/**
+ * Uses up up agression for mouse and simulates a fight between the given mouse
+ * and opponent it requests. Any mouse that loses in a fight dies.
+ * @param {Genetics.Mouse} mouse The mouse to initiate fighting.
+ */
+Genetics.Cage.instigateFight = function(mouse) {
+  var result = Genetics.Cage.runMouseFunction(mouse, 'pickFight');
+  // Check if function threw an error or caused a timeout.
+  if (result[0] == 'FAILURE') {
+    return;
+  }
+  var chosen = result[1];
+  // Check if no mouse was chosen.
+  if (chosen === null) {
+    new Genetics.Cage.Event('FIGHT', mouse.id, 'NONE').addToQueue();
+    mouse.aggressiveness = 0;
+    return;
+  }
+  // Check if return value is invalid.
+  if (typeof chosen != 'object' ||
+      Genetics.Cage.miceMap_[chosen.id] == undefined) {
+    new Genetics.Cage.Event('FIGHT', mouse.id, 'INVALID').addToQueue();
+    mouse.aggressiveness = 0;
+    return;
+  }
+  // Retrieve local copy of mouse to ensure that mouse was not modified.
+  var opponent = Genetics.Cage.miceMap_[chosen.id];
+  // Check if opponent is itself.
+  if (mouse.id == opponent.id) {
+    new Genetics.Cage.Event('FIGHT', mouse.id, 'SELF').addToQueue();
+    Genetics.Cage.death(mouse);
+    return;
+  }
+  // Compute winner using size to weigh probability of winning.
+  var fightResult = Math.random() - mouse.size / (mouse.size + opponent.size);
+  if (fightResult < 0) {  // If fight proposer won.
+    new Genetics.Cage.Event('FIGHT', mouse.id, 'WIN', opponent.id).addToQueue();
+    Genetics.Cage.death(opponent);
+  }
+  else if (fightResult == 0) {  // If it was a tie.
```

Comment:
@seanlip is correct.

The [original code](https://neil.fraser.name/scripts/moowebbrowser.py?db=MooCanada.db&obj=567&verbnum=2) was integer based.  So if mouseA was of size 6 and mouseB was of size 3, then random would be between 1 and 9 which beaks down to:

* 1-2 mouseB wins.  (22%)
* 3 tie.  (11%)
* 4-9 mouseA wins. (66%)

Thus the tie would happen quite frequently thanks to integer math.  Sorry that this was not obvious since most modern languages use float math.